### ISSUE TITLE
[Request] [Low Priority] Make likely/unlikely pragmas instead of normal templates

### ISSUE THREAD

**Issue by awr1** [nim-lang/Nim#8890]
Reccomend moving from something like this:
```nim
if likely(boolFunc()):
  # blah blah blah
elif unlikely(otherBoolFunc()):
  # blah blah blah
else:
  # blah blah blah
```
to something like this:
```nim
if boolFunc():
  {.likely.}
  # blah blah blah
elif otherBoolFunc():
  {.unlikely.}
  # blah blah blah
else:
  # blah blah blah
```
Likely/unlikely maps to the `__builtin_expect()` intrinsic in GCC/Clang to hint to the compiler a branch prediction consideration (if the target CPU uses branch predictor), of which there is no (presently existing) analogue on MSVC beyond profiling-based optimization (perhaps on MSVC, if we wanted to be fancy we could use a naive reordering strategy in the C codegen, i.e. put likely branches in the first branch if such an optimization is provably safe and does not break things). Considering we have [multiple other optimization-related pragmas](https://nim-lang.org/docs/manual.html#pragmas-linearscanend-pragma) regarding branching, I feel like this would make things more consistent.

**Comment by data-man** [comment]
Downvoted, sorry.

**Comment by dom96** [comment]
It's ``if (likely(foobar))`` in C too IIRC.

**Comment by Bulat-Ziganshin** [comment]
1. imho, indeed, pragmas should be used for things that doesn't change semantics. unfortunately, many Nim pragmas changes semantics
2. while likely pragma can help in some situations where likely function can't be used, there are situations when opposite holds: likely(a<0) and b>0

**Comment by Bulat-Ziganshin** [comment]
Yeah, i had a feeling that we have a lot of semantic-changing pragmas, and analyzing 0.18 docs I found the following:

> The final pragma can be used for an object type to specify that it cannot be inherited from.

In other languages it's a part of language. Adding this pragma breaks some programs

> An object type can be marked with the pure pragma so that its type field which is used for runtime type identification is omitted

This changes semantics. Types declared with and without pure are incompatible

> A proc can be marked with the asmNoStackFrame pragma to tell the compiler it should not generate a stack frame for the proc

This changes semantics. Procedures declared with and without it are incompatible

> callconv: Specifies the default calling convention for all procedures (and procedure types) that follow.

ditto

> The global pragma can be applied to a variable within a proc to instruct the compiler to store it in a global location

no comments

> The bitsize pragma is for object field members. It declares the field as a bitfield in C/C++
> The volatile pragma is for variables only. It declares the variable as  volatile

they are parts of C language for the reason.

> The incompleteStruct pragma tells the compiler to not use the underlying C struct in a sizeof expression

i.e. it changes sizeof semantics

Well, I think the bug in my definition. Pragmas definitely can change semantics - in particular, they are used for backend-specific functionality such as structure field size. And sometimes it seems they are used as temporary syntax for new features added to the language. Overall, pragmas look like "temporary" part of language that may be changed in the future, so you shouldn't rely on it unless absolutely required.

PS: https://nim-lang.org/docs/manual.html#typed-vs-untyped-parameters is referenced but not defined in the manual

**Comment by Araq** [comment]
> Well, I think the bug in my definition. Pragmas definitely can change semantics 

We can do without ``.final`` these days and well ... calling conventions and bit representations are not high level semantics. The C standard does not know ``__stdcall`` -- does that make it a better spec? I don't think so.

> Using likely to control jumps is almost irrelevant for modern CPUs, but I found another usage for it - "unlikely" code doesn't fight for registers, so by breaking code into main and "unlikely" parts, I can slightly control which variables will go to registers, making "likely" useful again.

Fair enough. Maybe. Seems better to name this "hot" and "cold" code then.

**Comment by Bulat-Ziganshin** [comment]
final, pure and global looks like backend-independent properties, so they may deserve real Nim keywords. Specifying them via pragmas imho should be considered as temporary solution until we made a final solution.
